vnext/Microsoft.ReactNative/CoalescingEventEmitter.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CoalescingEventEmitter.h"
+#include "CoalescingEventEmitter.g.cpp"
+#include "DynamicWriter.h"
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+/*static*/ ReactPropertyId<CoalescingEventEmitter> CoalescingEventEmitter::Property() noexcept {
+  static ReactPropertyId<CoalescingEventEmitter> propId{L"ReactNative.CoalescingEventEmitter", L"Emitter"};
+  return propId;
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    Mso::React::IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+CoalescingEventEmitter::CoalescingEventEmitter(Mso::CntPtr<Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {}
+
+void CoalescingEventEmitter::EmitJSEvent(int64_t coalesceKey, JSValueArgWriter const &paramsArgWriter) noexcept {
+  std::unique_lock lock(m_mutex);
+
+  for (auto &evt : m_eventQueue) {
+    if (evt.coalesceKey && *evt.coalesceKey == coalesceKey) {
+      evt.expired = true;
```

Comment:
When I thought deque was a circular buffer, I went this route bc it allows us to avoid shifting elements. Makes less sense for something linked or partially linked though.